### PR TITLE
Providing continuum x-ray emission

### PR DESCRIPTION
**Pull request by ehsteve**

### CODE REVIEW

File: roentgen/continuum/emission.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+__all__ = [
+    "thermal_bremsstrahlung",
+    "mewe_gaunt",
+    "xray_tube_spectrum",
+    "cutoff_wavelength"
+]
+
+
+def thermal_bremsstrahlung(wavelength, z, density, temperature):
+    """
```
Comment (hayesla) [inline roentgen/continuum/emission.py:15]: maybe a better description in the docs e.g. what is `z`?

File: roentgen/continuum/emission.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    This equation assumes that the density of ions is equal to the density of
+    electrons.
+    """
+    frequency = wavelength.to('Hz', equivalencies=u.spectral())
```
Comment (hayesla) [inline roentgen/continuum/emission.py:19]: I think this should be `to_value()` as `np.exp` doesn't take dimensional quantities.

File: roentgen/continuum/emission.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    return z ** 2 * density ** 2 * np.sqrt(1 / (k_B * temperature)) * \
+           np.exp(- h * frequency / (k_B * temperature)) * mewe_gaunt(wavelength, temperature)
+
+
+def mewe_gaunt(wavelength, temperature):
+    """Based on Equation 3 in Mewe 1985."""
+    return 27.83 * (temperature.to('MK').value + 0.65) ** (-1.33) + 0.15 * wavelength.to('angstrom').value ** (0.34) * temperature.to('MK').value ** (0.422)
```
Comment (hayesla) [inline roentgen/continuum/emission.py:26]: I'd say its best to `use to_value()`
`return 27.83 * (temperature.to_value('MK')+ 0.65) ** (-1.33) + 0.15 * wavelength.to_value('angstrom') ** (0.34) * temperature.to_value('MK') ** (0.422)`

Comment (hayesla) [reply roentgen/continuum/emission.py:26]: also similarly require that wavelength and temperature are astropy quantities with the correct units (`quantity_input`)

**Comment by DanRyanIrish** [comment]
Here is a plot comparing the output of ```thermal_bremsstrahlung()``` and the ```f_vth``` and ```brem_49``` SSW routines.  The ```roentgen``` curve has been scaled for easier comparison and the ```T```, ```EM```, ```n_e```, and ```z``` input parameters for each routine are listed in the plot title and legend.
image

File: roentgen/continuum/emission.py (hunk 2 above)
Comment (ehsteve) [reply roentgen/continuum/emission.py:19]: actually as long as the units cancel out in `np.exp` then it works fine!

Comment (hayesla) [reply roentgen/continuum/emission.py:19]: I think there is a 's' left over from the frequency?

**Comment by hayesla** [comment]
it looks like as @ehsteve mentioned that its difference in gaunt factor calculations - see figure below for comparison of gaunt factor calculation for roentgen `mewe_gaunt` and IDL `acgaunt`. The one here `mewe_gaunt` is the approximation where the IDL `acgaunt` uses the more full calculation  - 
 
mewe_comparison

File: roentgen/continuum/emission.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+
+def xray_tube_spectrum(wavelength, voltage):
```
[the remark this answers is not in the document]
Comment (ehsteve) [reply roentgen/continuum/emission.py:29]: No, just like the other spectrum, it only provides a relative spectrum. I was not able to find something that gave absolute strength of the spectrum. But good enough for the paper plots at this point.

**Comment by ehsteve** [comment]
@hayesla i'm confused because these two gaunt factors seem fairly close and so cannot explain the factor of ten difference I see in @DanRyanIrish at low energies but maybe this is a normalization issue? Dan seems to have normalized to match at higher energies so perhaps they deviate at high energies?

**Comment by ehsteve** [comment]
Could we get the plots on the same x range? say 1 to 200 keV?

**Comment by hayesla** [comment]
> Could we get the plots on the same x range? say 1 to 200 keV?

oh actually I was playing around with it, and if you scale the output by energy (i.e. keV) and set the density to 0.01 it matches pretty well with the offsets more in line with the gaunt factors. I'm a bit confused @ehsteve - what is the density units supposed to be in `thermal_bremsstrahlung`?

both_comparsion_fixed

**Comment by ehsteve** [comment]
That's great! Thanks @hayesla. The density value is `n_i * n_e`. For solar we usually integrate that out into the emission measure. I just wanted to capture the dependencies but did not have the time to calculate the correct constant (which is why all the units are messed up as well). Alright, to me this is good enough. Certainly for the attenuator paper where we only need the shape to be correct. What do you think @DanRyanIrish?

**Comment by ehsteve** [comment]
@DanRyanIrish could you compare the XrayTubeSpectrum provided here with the spectra provided by the website you found http://solutioinsilico.com/medical-physics/applications/tasmip-app.php?ans=1&n_sp=1&k1=30&m1=1&t1=100&k2=&m2=1&t2=&k3=&m3=1&t3=

File: roentgen/continuum/emission.py (hunk 3 above)
Comment (DanRyanIrish) [reply roentgen/continuum/emission.py:67]: I believe this line still needs the equivalency added: ```wavelength.to_value('angstrom', equivalencies=u.spectral())```

File: roentgen/continuum/emission.py [hunk 5]
```diff
@@ -4,40 +4,109 @@
...
 
 __all__ = [
-    "thermal_bremsstrahlung",
-    "mewe_gaunt",
-    "xray_tube_spectrum",
-    "cutoff_wavelength"
+    "ThermalBremsstrahlung",
+    "GauntFactor",
+    "XraytubeEmission",
+    "CutoffWavelength"
```
Comment (DanRyanIrish) [inline roentgen/continuum/emission.py:10]: Is there a reason these functions are named with camel case, like classes?

**Comment by DanRyanIrish** [comment]
```XraytubeSpectrum``` fails when energy (keV) given rather than wavelength (nm).  May be an issue with ```CutoffWavelength```

```python
In [18]: roentgen_data = XraytubeEmission(web_data["energy"], 100*u.kV)                                                                                          
---------------------------------------------------------------------------
UnitConversionError                       Traceback (most recent call last)
<ipython-input-18-f72e14b5789e> in <module>
----> 1 roentgen_data = XraytubeEmission(web_data["energy"], 100*u.kV)

~/miniconda3/envs/roentgen-dev/lib/python3.7/site-packages/astropy/units/decorators.py in wrapper(*func_args, **func_kwargs)
    232             # Call the original function with any equivalencies in force.
    233             with add_enabled_equivalencies(self.equivalencies):
--> 234                 return_ = wrapped_function(*func_args, **func_kwargs)
    235             if wrapped_signature.return_annotation not in (inspect.Signature.empty, None):
    236                 return return_.to(wrapped_signature.return_annotation)

~/sunpy_dev/roentgen/roentgen/continuum/emission.py in XraytubeEmission(wavelength, voltage)
     85     emission : The unscaled emission, provided without units
     86     """
---> 87     f = (wavelength / CutoffWavelength(voltage) - 1) * 1 / wavelength ** 2
     88     # remove unphysical negative values
     89     f[f < 0.0] = 0.0

~/miniconda3/envs/roentgen-dev/lib/python3.7/site-packages/astropy/units/quantity.py in __array_ufunc__(self, function, method, *inputs, **kwargs)
    455         # consistent units between two inputs (e.g., in np.add) --
    456         # and the unit of the result (or tuple of units for nout > 1).
--> 457         converters, unit = converters_and_unit(function, method, *inputs)
    458 
    459         out = kwargs.get('out', None)

~/miniconda3/envs/roentgen-dev/lib/python3.7/site-packages/astropy/units/quantity_helper/converters.py in converters_and_unit(function, method, *args)
    187                             "argument is not a quantity (unless the "
    188                             "latter is all zero/infinity/nan)"
--> 189                             .format(function.__name__))
    190             except TypeError:
    191                 # _can_have_arbitrary_unit failed: arg could not be compared

UnitConversionError: Can only apply 'subtract' function to dimensionless quantities when other argument is not a quantity (unless the latter is all zero/infinity/nan)
```

**Comment by DanRyanIrish** [comment]
Here is the comparison of the roentgen and web-tool x-ray tube spectra for a tube voltage of 100 kV
image

**Comment by ehsteve** [comment]
Other packages are to provide this. Out of scope.

**Pull request closed without merging**